### PR TITLE
STY: remove --keep-runtime-typing from pyupgrade Part-1

### LINKED ISSUE #40759: STYLE remove --keep-runtime-typing from pyupgrade
xref @simonjayhawkins 's comment: https://github.com/pandas-dev/pandas/pull/40545#discussion_r606360410

Task is:
- remove `keep-runtime-typing` from:

https://github.com/pandas-dev/pandas/blob/cc84a2383c641998d4817cfaeb33c64f017b7454/.pre-commit-config.yaml#L57

- run `pre-commit run pyupgrade --all-files`
- run `pre-commit run flake8 --all-files`, fixup any errors (probably some about extra imports)

Finally, run `pre-commit run --all-files` (this may take a few minutes), fixup any hooks which don't pass, and when everything passes, open a pull request :rocket:

### PR DESCRIPTION
**Pull request by fathomer**
- [x] xref #40759
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
The mypy failures are due to the fact that some of these methods have their own method `type`, which interferes with the built-in `type`

small example:
```python
# t.py

from __future__ import annotations
from typing import Type


class Cat:
    
    def type(self) -> str:
        return 'this is my type'

    def type_cat(self) -> Type[Cat]:
        return Cat
```

works fine:
```console
$ mypy t.py 
Success: no issues found in 1 source file
```
but the following:
```python
# t.py

from __future__ import annotations
from typing import Type


class Cat:
    
    def type(self) -> str:
        return 'this is my type'

    def type_cat(self) -> type[Cat]:
        return Cat
```
doesn't:
```console
$ mypy t.py 
t.py:10: error: Function "t.Cat.type" is not valid as a type
t.py:10: note: Perhaps you need "Callable[...]" or a callback protocol?
Found 1 error in 1 file (checked 1 source file)
```
If we moved the definitions around so that `type_cat` came before `type`, it would work.

---

I think the solution here is to define the alias `type_t = type` in `pandas/_typing.py`, and then import that, and use it here. This seems analogous to using `bool_t` in `pandas/core/generic.py` (#40175)

cc @WillAyd @simonjayhawkins

**Comment by fathomer** [comment]
Thanks @MarcoGorelli for the nice explanation and suggestion, I have made changes according to that and pushed my code.

**Comment by MarcoGorelli** [comment]
Nice, thanks @fathomer ! Let's wait to hear from the others about this `import Type as Type_t` workaround

File: pandas/_typing.py
Comment (jreback) [inline pandas/_typing.py:28]: why not TypeT

Comment (jreback) [reply pandas/_typing.py:28]: or type_t ?

**Review by jreback — CHANGES_REQUESTED** [review]
this will get quickly out of date. i would rebase now and we'll see if cna merge it. but maybe better to break into pieces.

confirming this is basically List -> list, Dict -> dict, Tuple -> tuple (can you put the list in the top).

if its more complicated than that we should do in steps.

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Agree on breaking into smaller pieces - if you apply these changes to, say, 15-20 files at a time, then the current check will still pass even if it has `--keep-runtime-typing`.

If you could break this into a few smaller pull requests, each with no more than 20 files, then we can get those in, and at the end, we can remove `--keep-runtime-typing`

---

Jeff - the changes here are from PEP585 (e.g. `List[str]` -> `list[str]`) and from PEP604 (`Optional[str]` -> `str | None`), both of which are only applied in the presence of `from __future__ import annotations`

File: pandas/_typing.py
Comment (MarcoGorelli) [reply pandas/_typing.py:28]: you're right, `type_t` would be more consistent

Comment (fathomer) [reply pandas/_typing.py:28]: So I will change the variable name to type_t, keep the changes to 15-20 files as of now, and still keep the --keep-runtime-typing flag there in precommit hook.
Is that right @MarcoGorelli ?

Comment (MarcoGorelli) [reply pandas/_typing.py:28]: yes, that's right @fathomer! Then, in the final PR, you can remove the flag

Comment (fathomer) [reply pandas/_typing.py]: I have reverted changes to all other files except some 20 files for this PR, did a rebase to upstream/master and pushed.
For the remaining files do you want me raise the other PRs just now or wait until this one gets merged ?

Comment (MarcoGorelli) [reply pandas/_typing.py]: Thanks @fathomer! You can raise the others now if you want, I'll look over them tomorrow

Comment (fathomer) [reply pandas/_typing.py]: @MarcoGorelli @jreback Raised the other PRs as well, for some reason not getting the option to add you guys as reviewers there. Please feel free to review them.

**Comment by MarcoGorelli** [comment]
Looks good, thanks!

**Pull request merged**
